### DISCUSSION TITLE
How can I get a binding value in MarkupExtension??

### DISCUSSION THREAD

**Discussion by jim-jiang-github** [AvaloniaUI/Avalonia#11401] [Q&A] [answered]
Like the code:
![image](https://github.com/AvaloniaUI/Avalonia/assets/29189410/d83e44f5-93d3-4c65-9a3c-bd4a94539421)

I want to get the Binding value 'Text="{i18n:Localize {Binding Version}, Key=Test}"' '{Binding Version}', the binding value may be binding from viewmodel or self property, and we can get the 'Binding' object in MarkupExtension.ProvideValue, But how can we get the binding value???
In wpf we can do like 'var resolvedValue = b.ProvideValue(new ServiceProvider(dependencyObject, dependencyProperty));'.

**Comment by thexamlguy** [comment]
I believe something like below should be sufficient enough... 

An attached property to act as a binding proxy somewhere within the class...

`private static readonly AvaloniaProperty ValueProperty =
            AvaloniaProperty.RegisterAttached<LocalizeExtension, Control, object>("Value");`
            
   and then in your ProvideValue method...
   
   	if (arg is Binding binding)
            {
                targetObject.Bind(ValueProperty, binding);
                if (targetObject.GetValue(ValueProperty) is object value)
                {
                    // the value should be your actual value
                }
            }

**Reply by jim-jiang-github** [reply]
> I believe something like below should be sufficient enough...
> 
> An attached property to act as a binding proxy somewhere within the class...
> 
> `private static readonly AvaloniaProperty ValueProperty = AvaloniaProperty.RegisterAttached<LocalizeExtension, Control, object>("Value");`
> 
> and then in your ProvideValue method...
> 
> ```
> if (arg is Binding binding)
>         {
>             targetObject.Bind(ValueProperty, binding);
>             if (targetObject.GetValue(ValueProperty) is object value)
>             {
>                 // the value should be your actual value
>             }
>         }
> ```

![image](https://github.com/AvaloniaUI/Avalonia/assets/29189410/0316f83c-4f21-46bd-8f8c-8eb53b6b80e3)

It seems  no work? The value is 'null'

Here is my demo repo: https://github.com/jim-jiang-github/WpfLocalization/tree/avalonia
Branch: avalonia

**Reply by maxkatz6** [reply, marked as answer]
@jim-jiang-github since you have everything - both target object and property - in the extension, you can easily instanciate the binding:
```csharp
if (arg is Binding b)
{
    var instancedBinding = b.Initiate(targetObject, targetProperty);
    // If Observable is null, then it probably was OneWayToSource binding,
    // and it's not possible to read. 
    if (instancedBinding?.Observable is { } observable)
    {
        observable.Subscribe(s =>
        {

        });
    }
}
```

![image](https://github.com/AvaloniaUI/Avalonia/assets/3163374/cce53a1d-6843-426b-853a-93332a631da0)

**Reply by maxkatz6** [reply]
Note, that also means, that you need to change your code with BindingData to accept observables. Without that you won't be able to receive delated or multiple values updates from the argument-binding.

**Reply by maxkatz6** [reply]
If you don't care about binding updates, and want to read only current value, then you can try this hack - https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Primitives/SelectingItemsControl.cs#L1371
I.e., creating a fake property and object to receive this value.
But I wouldn't recommend that.

**Reply by maxkatz6** [reply]
Keep in mind, that if you also can implement your own IBinding class and return it from the markup extension. Instead of using ReflectionBinding hack from your code.
Our DynamicResource is a good example of that: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Markup/Avalonia.Markup.Xaml/MarkupExtensions/DynamicResourceExtension.cs

It's not mentioned in the documentation, as it's more an advanced use case. But you can learn a lot from the code.

**Reply by jim-jiang-github** [reply]
Thank you very much for your answer. I have been bothered by this question for several days.

**Reply by jl0pd** [reply]
@maxkatz6 `BindingBase.Initiate` is obsolete now, custom bindings cannot be created because of `IBinding2`. What is recommended way?

**Comment by Coloryr** [comment]
I have more better `i18n` funtion.
https://github.com/Coloryr/ColorMC/blob/master/src/ColorMC.Gui/Utils/LaunchSetting/Localizer.cs
https://github.com/Coloryr/ColorMC/blob/master/src/ColorMC.Gui/Utils/LaunchSetting/Extensions.cs
```
<Label
                    Margin="20,0,0,0"
                    HorizontalAlignment="Left"
                    VerticalAlignment="Center"
                    Content="{setting:Localize AddModPackWindow.Control.Text1}" />
```

**Reply by jim-jiang-github** [reply]
Your approach is not bad, but it cannot achieve parameterized localization, for example, this kind of effect. I would prefer a more comprehensive functionality.

![image](https://github.com/AvaloniaUI/Avalonia/assets/29189410/f0e0265c-a216-4ea8-8224-7c3f9bc89c14)
![image](https://github.com/AvaloniaUI/Avalonia/assets/29189410/7e5584ac-28ed-4237-9c52-0f2cd61561f1)

**Reply by BuriKizilkaya** [reply]
Have you solved it and how? Can you show me some example code for the custom IBinding Interface?

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/Avalonia/discussions/13448#discussioncomment-7699230
